File: backend/app/models/portaluser.py
from ..db import db
from sqlalchemy import UUID
import enum


class StatusEnum(enum.Enum):
    active = 'active'
    deleted = 'deleted'
    pending = 'pending'


class PortalUser(db.Model):
    __tablename__ = 'portalusers'
    guid = db.Column(UUID, primary_key=True, nullable=False, unique=True)
    userid = db.Column(db.String(120), nullable=False, unique=True)
    username = db.Column(db.String(120), nullable=False)
    lastname = db.Column(db.String(120), nullable=False)
    firstname = db.Column(db.String(120), nullable=False)
    fullname = db.Column(db.String(120), nullable=False)
    email = db.Column(db.String(120), nullable=False)
    homepage = db.Column(db.String(120), nullable=False)
    description = db.Column(db.String(255), nullable=True)
    status = db.Column(db.Enum(StatusEnum), nullable=False, default=StatusEnum.active)
    lastlogin = db.Column(db.DateTime, nullable=True)
    modified = db.Column(db.DateTime, nullable=False)
    created = db.Column(db.DateTime, nullable=False)
    provider = db.Column(db.String(120), nullable=False)
    role = db.Column(db.String(120), nullable=False)
    roleid = db.Column(db.String(120), nullable=False)
    customrole = db.Column(db.String(120), nullable=False)
    disabled = db.Column(db.Boolean, nullable=False)
    licensetype = db.Column(db.String(120), nullable=False)
    usertype = db.Column(db.String(120), nullable=False)
    access = db.Column(db.String(120), nullable=False)
    storeage = db.Column(db.Integer, nullable=False)
    itemcount = db.Column(db.Integer, nullable=False)
    groupcount = db.Column(db.Integer, nullable=False)
    adstatus = db.Column(db.String(120), nullable=False)
    department = db.Column(db.String(120), nullable=False)
    agency = db.Column(db.String(120), nullable=False)
    division = db.Column(db.String(120), nullable=False)
    team = db.Column(db.String(120), nullable=False)

    tasks = db.relationship(
        'Task',
        primaryjoin="and_(foreign(Task.linked_object_guid) == PortalUser.guid, Task.linked_object_type == 'PortalUser')"
    )


    def to_dict(self):
        return {
            "guid": self.guid,
            "userid": self.userid,
            "username": self.username,
            "lastname": self.lastname,
            "firstname": self.firstname,
            "fullname": self.fullname,
            "email": self.email,
            "homepage": self.homepage,
            "description": self.description,
            "status": self.status.name if self.status else None,
            "lastlogin": self.lastlogin,
            "modified": self.modified,
            "created": self.created,
            "provider": self.provider,
            "role": self.role,
            "roleid": self.roleid,
            "customrole": self.customrole,
            "disabled": self.disabled,
            "licensetype": self.licensetype,
            "usertype": self.usertype,
            "access": self.access,
            "storeage": self.storeage,
            "itemcount": self.itemcount,
            "groupcount": self.groupcount,
            "adstatus": self.adstatus,
            "department": self.department,
            "agency": self.agency,
            "division": self.division,
            "tean": self.team
        }

    def __repr__(self):
        return f"<Username {self.username}>"

File: frontend/src/models/portaluser.ts
import {AbstractModel} from "./abstract-models.ts";
import {makeObservable, observable} from 'mobx';

export enum StatusEnum {
    Active = 'active',
    Deleted = 'deleted',
    Pending = 'pending'
}

export interface PortalUserProps {
    guid: string;
    userid: string;
    username: string;
    lastname: string;
    firstname: string;
    fullname: string;
    email: string;
    homepage: string;
    description?: string;
    status: StatusEnum;
    lastlogin?: Date;
    modified: Date;
    created: Date;
    provider: string;
    role: string;
    roleid: string;
    customrole: string;
    disabled: boolean;
    licensetype: string;
    usertype: string;
    access: string;
    storeage: number;
    itemcount: number;
    groupcount: number;
    adstatus: string;
    division1: string;
    division2: string;
    division3: string;
    division4: string;
}

export class PortalUser extends AbstractModel {
    class: string = 'PortalUser';

    guid: string;
    userid: string;
    username: string;
    lastname: string;
    firstname: string;
    fullname: string;
    email: string;
    homepage: string;
    description?: string;
    status: StatusEnum;
    lastlogin?: Date;
    modified: Date;
    created: Date;
    provider: string;
    role: string;
    roleid: string;
    customrole: string;
    disabled: boolean;
    licensetype: string;
    usertype: string;
    access: string;
    storeage: number;
    itemcount: number;
    groupcount: number;
    adstatus: string;
    division1: string;
    division2: string;
    division3: string;
    division4: string;

    constructor(props: PortalUserProps) {
        super()
        this.guid = props.guid;
        this.userid = props.userid;
        this.username = props.username;
        this.lastname = props.lastname;
        this.firstname = props.firstname;
        this.fullname = props.fullname;
        this.email = props.email;
        this.homepage = props.homepage;
        this.description = props.description;
        this.status = props.status;
        this.lastlogin = props.lastlogin;
        this.modified = props.modified;
        this.created = props.created;
        this.provider = props.provider;
        this.role = props.role;
        this.roleid = props.roleid;
        this.customrole = props.customrole;
        this.disabled = props.disabled;
        this.licensetype = props.licensetype;
        this.usertype = props.usertype;
        this.access = props.access;
        this.storeage = props.storeage;
        this.itemcount = props.itemcount;
        this.groupcount = props.groupcount;
        this.adstatus = props.adstatus;
        this.division1 = props.division1;
        this.division2 = props.division2;
        this.division3 = props.division3;
        this.division4 = props.division4;
    }
}

// export class PortalUser extends AbstractModel {
//     class: string = 'PortalUser';
//     guid: string = '';
//     userid: string = '';
//     username: string = '';
//     lastname: string = '';
//     firstname: string = '';
//     fullname: string = '';
//     email: string = '';
//     homepage: string = '';
//     description?: string;
//     status: StatusEnum = StatusEnum.Pending;
//     lastlogin?: Date;
//     modified: Date = new Date();
//     created: Date = new Date();
//     provider: string = '';
//     role: string = '';
//     roleid: string = '';
//     customrole: string = '';
//     disabled: boolean = false;
//     licensetype: string = '';
//     usertype: string = '';
//     access: string = '';
//     storeage: number = 0;
//     itemcount: number = 0;
//     groupcount: number = 0;
//     adstatus: string = '';
//     division1: string = '';
//     division2: string = '';
//     division3: string = '';
//     division4: string = '';
//
//     constructor(props?: PortalUserProps) {
//         super();
//         if (props) {
//             Object.assign(this, props);
//
//         }
//         makeObservable<this, 'guid' | 'userid' | 'username' | 'lastname' | 'firstname' | 'fullname' | 'email' | 'homepage' | 'description' | 'status' | 'lastlogin' | 'modified' | 'created' | 'provider' | 'role' | 'roleid' | 'customrole' | 'disabled' | 'licensetype' | 'usertype' | 'access' | 'storeage' | 'itemcount' | 'groupcount' | 'adstatus' | 'division1' | 'division2' | 'division3' | 'division4'>(this, {
//             guid: observable,
//             userid: observable,
//             username: observable,
//             lastname: observable,
//             firstname: observable,
//             fullname: observable,
//             email: observable,
//             homepage: observable,
//             description: observable,
//             status: observable,
//             lastlogin: observable,
//             modified: observable,
//             created: observable,
//             provider: observable,
//             role: observable,
//             roleid: observable,
//             customrole: observable,
//             disabled: observable,
//             licensetype: observable,
//             usertype: observable,
//             access: observable,
//             storeage: observable,
//             itemcount: observable,
//             groupcount: observable,
//             adstatus: observable,
//             division1: observable,
//             division2: observable,
//             division3: observable,
//             division4: observable
//         });
//     }
// }